src/main/java/com/mraof/minestuck/player/Title.java

```diff
@@ -1,79 +1,128 @@
 	{
-		this.heroClass = Objects.requireNonNull(heroClass);
-		this.heroAspect = Objects.requireNonNull(heroAspect);
+		return getTitle(player)
+				.map(title -> title.heroAspect() == aspect)
+				.orElse(false);
 	}
 	
-	public EnumClass getHeroClass()
+	public static Optional<Title> getTitle(ServerPlayer player)
 	{
-		return this.heroClass;
+		return PlayerData.get(player).flatMap(Title::getTitle);
 	}
 	
-	public EnumAspect getHeroAspect()
+	public static Optional<Title> getTitle(PlayerIdentifier playerId, MinecraftServer mcServer)
 	{
-		return this.heroAspect;
+		return getTitle(PlayerData.get(playerId, mcServer));
+	}
+	
+	public static Optional<Title> getTitle(PlayerData playerData)
+	{
+		return playerData.getExistingData(MSAttachments.TITLE);
+	}
+	
+	public static void setTitle(PlayerData playerData, Title newTitle)
+	{
+		if(getTitle(playerData).isPresent())
+			throw new IllegalStateException("Can't set title for player " + playerData.identifier.getUsername() + " because they already have one");
+		
+		playerData.setData(MSAttachments.TITLE, newTitle);
+		
+		ServerPlayer player = playerData.getPlayer();
+		if(player != null)
+			player.connection.send(TitleDataPacket.create(newTitle));
 	}
 	
 	public void handleAspectEffects(ServerPlayer player)
 	{
-		if(!MinestuckConfig.SERVER.aspectEffects.get())
+		if(!MinestuckConfig.SERVER.aspectEffects.get() || !player.getData(MSAttachments.EFFECT_TOGGLE))
+			return;
+		if(player.getCommandSenderWorld().getGameTime() % 380 != 0)
 			return;
-		PlayerData data = PlayerSavedData.getData(player);
-		if(data.effectToggle())
+		Optional<PlayerData> data = PlayerData.get(player);
+		if(data.isEmpty())
+			return;
+		
+		int rung = Echeladder.get(data.get()).getRung();
+		EnumAspect aspect = this.heroAspect();
+		int potionLevel = (int) (ASPECT_EFFECTS.get(aspect).strength() * rung);
+		
+		if(rung > 18 && aspect == EnumAspect.HOPE)
+			player.addEffect(new MobEffectInstance(MobEffects.WATER_BREATHING, 600, 0));
+		
+		if(potionLevel > 0)
 		{
-			int rung = data.getEcheladder().getRung();
-			EnumAspect aspect = data.getTitle().getHeroAspect();
-			int potionLevel = (int) (aspectStrength[aspect.ordinal()] * rung); //Blood, Breath, Doom, Heart, Hope, Life, Light, Mind, Rage, Space, Time, Void
-			
-			if(player.getCommandSenderWorld().getGameTime() % 380 == 0)
-			{
-				if(rung > 18 && aspect == HOPE)
-				{
-					player.addEffect(new MobEffectInstance(MobEffects.WATER_BREATHING, 600, 0));
-				}
-				
-				if(potionLevel > 0)
-				{
-					player.addEffect(new MobEffectInstance(aspectEffects[aspect.ordinal()], 600, potionLevel - 1));
-					LOGGER.debug("Applied aspect potion effect to {}", player.getDisplayName().getString());
-				}
-			}
+			player.addEffect(new MobEffectInstance(ASPECT_EFFECTS.get(aspect).effect(), 600, potionLevel - 1));
+			LOGGER.debug("Applied aspect potion effect to {}", player.getDisplayName().getString());
```

Comment:
The log message applies twice at the same time, making me think the effect is also being applied twice

Reply:
That might be `ServerEventHandler` listening to the player tick event and handling both the start and end phase of the tick.